### PR TITLE
Introduce `Stateful` DataModule

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

part of #11429: introduce `Stateful` DataModule (see **Appendix** code changes outlined in issue)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_datamodules.py [hunk 1]
```diff
@@ -221,13 +276,14 @@ def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
...
+    assert "my_dm_state" in checkpoint
+    assert checkpoint["my_dm_state"] == {"my": "state_dict"}
 
     for trainer_fn in TrainerFn:
         trainer.state.fn = trainer_fn
-        with mock.patch.object(dm, "on_load_checkpoint") as dm_mock:
-            trainer._restore_modules_and_callbacks(checkpoint_path)
-            dm_mock.assert_called_once()
+        trainer._restore_modules_and_callbacks(checkpoint_path)
+        assert dm.my_state_dict == {"my": "state_dict", "on_load": "update"}
```
Comment (jjenniferdai) [inline tests/core/test_datamodules.py:286]: example failure - because we introduce `load_state_dict` call after `on_load_checkpoint` - if a user does already have a `Stateful` datamodule (and does not expect lightning to call `load_state_dict`), changes made in `on_load_checkpoint` could be overridden by lightning's `load_state_dict` call after it.

What's the smoothest way to resolve/introduce this change?

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -257,3 +257,9 @@ def test_dataloader():
         if test_dataset is not None:
             datamodule.test_dataloader = test_dataloader
         return datamodule
+
+    def state_dict(self) -> Dict[str, Any]:
+        return {}
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        pass
```
Comment (ananthsub) [inline pytorch_lightning/core/datamodule.py:265]: please add docstrings

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -385,10 +387,15 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
             else:
                 checkpoint[pl.LightningModule.CHECKPOINT_HYPER_PARAMS_KEY] = dict(model.hparams)
 
-        # give the model a chance to dump a few things
+        # dump stateful datamodule
+        datamodule = self.trainer.datamodule
+        if datamodule is not None:
+            checkpoint[datamodule.__class__.__name__] = datamodule.state_dict()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:393]: all `state_dict()` calls should happen before any `on_save_checkpoint` hooks are called

File: tests/core/test_datamodules.py (hunk 1 above)
Comment (ananthsub) [reply tests/core/test_datamodules.py:286]: taken a step further, the user may have defined a load_state_dict/state_dict with different arguments, and the lightning trainer calling this function will result in an error because of a mismatched signature. 

I don't think there's any smooth way to get around this because it's a risk of adding to the superclass and relying on inheritance/OOP

however i think the chance of this happening is very, very small.

**Review by ananthsub — COMMENTED** [review]
as a follow up, please deprecate CheckpointHooks off the PrecisionPlugin interface: https://github.com/PyTorchLightning/pytorch-lightning/blob/86b177ebe5427725b35fde1a8808a7b59b8a277a/pytorch_lightning/core/datamodule.py#L26

It is no longer needed following the Stateful change and should be removed in v1.8

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -154,6 +154,8 @@ def restore_datamodule(self) -> None:
         datamodule = self.trainer.datamodule
         if datamodule is not None:
             datamodule.on_load_checkpoint(self._loaded_checkpoint)
+            if datamodule.__class__.__name__ in self._loaded_checkpoint:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:157]: https://github.com/PyTorchLightning/pytorch-lightning/blob/86b177ebe5427725b35fde1a8808a7b59b8a277a/pytorch_lightning/callbacks/base.py#L43 

let's use qualname here too for consistency

**Comment by jjenniferdai** [comment]
Did we want to deprecate `CheckpointHooks` off `DataModule`? The original issue #11429 discussion was more focused only on adding the `Stateful` alignment for all components (Motivation point #1) without editing `CheckpointHooks` (Motivation point #2) yet

**Comment by ananthsub** [comment]
> Did we want to deprecate `CheckpointHooks` off `DataModule`? The original issue #11429 discussion was more focused only on adding the `Stateful` alignment for all components (Motivation point 1) without editing `CheckpointHooks` (Motivation point 2) yet

I should have clarified that we can follow it up in a separate issue/PR. It is not a blocker for this one.

I initially proposed using the CheckpointHooks to DataModule back in https://github.com/PyTorchLightning/pytorch-lightning/issues/3544 . It was to capture dataloader state that had proprietary APIs. I wish at the time I had recognized the `Stateful` protocol lurking in pytorch and the rest of the lightning codebase as the contract for saving/loading state, but I didn't. I don't think there's a use case for the datamodule to be augmenting/changing other components state through the `on_save/load_checkpoint` hooks.

**Review by ananthsub — APPROVED** [review]
nice!

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -331,7 +333,7 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
                 CHECKPOINT_HYPER_PARAMS_KEY:
                 CHECKPOINT_HYPER_PARAMS_TYPE:
                 something_cool_i_want_to_save: anything you define through model.on_save_checkpoint
-                LightningDataModule.__class__.__name__: pl DataModule's state
+                LightningDataModule.__class__.__qualname__: pl DataModule's state
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:336]: I strongly prefer the `__qualname__` with a deprecation-phase where we support both. This is way more unique. And unless that class is not defined in the global scope of a file (which would be a problem with spawn etc. anyways, they should be identical

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/datamodule.py [hunk 6]
```diff
@@ -257,3 +257,19 @@ def test_dataloader():
...
+
+    def state_dict(self) -> Dict[str, Any]:
+        """Called when saving a checkpoint, implement to generate and save datamodule state.
+
+        Returns:
+            A dictionary containing datamodule state.
+        """
+        return {}
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
```
Comment (rohitgr7) [inline pytorch_lightning/core/datamodule.py:269]:
```suggestion
    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
```
quick ques: does `Dist` supports `OrderedDict`? since, for PT Module, it is OrderedDict. Or should we use `Mapping`

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:336]: I don't think the old checkpoint format was actually `LightningDataModule.__class__.__name__` ?`LightningDataModule` could just update/add to the dict with `on_save_checkpoint` but lightning never dictated `LightningDataModule.__class__.__name__` key

File: pytorch_lightning/core/datamodule.py (hunk 6 above)
Comment (jjenniferdai) [reply pytorch_lightning/core/datamodule.py:269]: yep checked mypy has no issues with this:

```
def testingmethod() -> Dict[Any, Any]:
    return OrderedDict()
```
